File: shared/src/text/numbers.ts
const UNITS: Readonly<Record<string, number>> = {
  zéro: 0, zero: 0, un: 1, une: 1, deux: 2, trois: 3, quatre: 4, cinq: 5, six: 6, sept: 7, huit: 8, neuf: 9,
  dix: 10, onze: 11, douze: 12, treize: 13, quatorze: 14, quinze: 15, seize: 16,
};
const TENS: Readonly<Record<string, number>> = {
  vingt: 20, vingts: 20, trente: 30, quarante: 40, cinquante: 50, soixante: 60, septante: 70, octante: 80, huitante: 80, nonante: 90,
};
const HUNDREDS = new Set(['cent', 'cents']);
const SCALES: Readonly<Record<string, number>> = { mille: 1000, mil: 1000, million: 1e6, millions: 1e6, milliard: 1e9, milliards: 1e9 };

/** Every word that may appear inside a French number written in letters. */
export const NUMBER_WORDS: ReadonlySet<string> = new Set([...Object.keys(UNITS), ...Object.keys(TENS), ...HUNDREDS, ...Object.keys(SCALES)]);

const ORDINAL_SUFFIX = '(?:e|es|er|ers|re|res|ère|ères|ème|èmes|eme|emes|è|nd|nde|nds|ndes)';
const DIGITS_RE = new RegExp(
  `^([+\\-\u2212]?)(\\d{1,3}(?:[ .]\\d{3})+|\\d+)(?:[,.](\\d+))?(?:(${ORDINAL_SUFFIX}))?(?: (mille|millions?|milliards?))?$`,
);
const ROMAN_RE = new RegExp(`^(?=[IVXLCDM])(M{0,3})(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})(${ORDINAL_SUFFIX})?$`);
const ROMAN_VALUES: Readonly<Record<string, number>> = { I: 1, V: 5, X: 10, L: 50, C: 100, D: 500, M: 1000 };

type Parsed = [value: number, next: number];

function unitIn(word: string | undefined, min: number, max: number): number | null {
  if (word === undefined) return null;
  const v = UNITS[word];
  return v !== undefined && v >= min && v <= max ? v : null;
}

/** 1..19 after « soixante » / « quatre-vingt » (teens allowed) or 1..9 after other tens. */
function parseTail(w: string[], i: number, allowTeens: boolean): Parsed | null {
  if (allowTeens) {
    if (w[i] === 'dix') {
      const u = unitIn(w[i + 1], 7, 9);
      return u !== null ? [10 + u, i + 2] : [10, i + 1];
    }
    const teen = unitIn(w[i], 11, 16);
    if (teen !== null) return [teen, i + 1];
  }
  const u = unitIn(w[i], 1, 9);
  return u !== null ? [u, i + 1] : null;
}

function parseBelow100(w: string[], i: number): Parsed | null {
  const a = w[i];
  if (a === undefined) return null;
  if (a === 'quatre' && (w[i + 1] === 'vingt' || w[i + 1] === 'vingts')) {
    const tail = w[i + 1] === 'vingt' ? parseTail(w, i + 2, true) : null;
    return tail ? [80 + tail[0], tail[1]] : [80, i + 2];
  }
  const ten = TENS[a];
  if (ten !== undefined) {
    const tail = parseTail(w, i + 1, ten === 60);
    return tail ? [ten + tail[0], tail[1]] : [ten, i + 1];
  }
  if (a === 'dix') {
    const u = unitIn(w[i + 1], 7, 9);
    if (u !== null) return [10 + u, i + 2];
  }
  const unit = UNITS[a];
  return unit !== undefined ? [unit, i + 1] : null;
}

function parseBelow1000(w: string[], i: number): Parsed | null {
  let hundreds = 0;
  let j = i;
  if (HUNDREDS.has(w[j] ?? '')) {
    hundreds = 100;
    j++;
  } else {
    const u = unitIn(w[j], 2, 9);
    if (u !== null && HUNDREDS.has(w[j + 1] ?? '')) {
      hundreds = u * 100;
      j += 2;
    }
  }
  if (hundreds > 0 && w[j] !== undefined && (w[j] === 'zéro' || w[j] === 'zero')) return [hundreds, j];
  const rest = parseBelow100(w, j);
  if (rest) {
    if (hundreds > 0 && rest[0] === 0) return [hundreds, j];
    return [hundreds + rest[0], rest[1]];
  }
  return hundreds > 0 ? [hundreds, j] : null;
}

/** Value of a whole list of cardinal number words, or null when the list is not exactly one number. */
function parseCardinal(w: string[]): number | null {
  if (w.length === 0) return null;
  if (w.length === 1 && (w[0] === 'zéro' || w[0] === 'zero')) return 0;
  let total = 0;
  let lastScale = Number.POSITIVE_INFINITY;
  let i = 0;
  while (i < w.length) {
    const group = parseBelow1000(w, i);
    if (group && group[0] === 0) return null;
    const next = group ? group[1] : i;
    const scaleWord = w[next];
    const scale = scaleWord !== undefined ? SCALES[scaleWord] : undefined;
    if (scale !== undefined) {
      if (scale >= lastScale) return null;
      if (!group && scale !== 1000) return null;
      total += (group ? group[0] : 1) * scale;
      lastScale = scale;
      i = next + 1;
      continue;
    }
    if (!group || next !== w.length) return null;
    return total + group[0];
  }
  return total;
}

function splitNumberWords(s: string): string[] {
  return s
    .toLowerCase()
    .split(/[\s\-\u2010\u2011]+/)
    .filter((x) => x.length > 0 && x !== 'et');
}

const ORDINAL_STEMS: Readonly<Record<string, string>> = { cinqu: 'cinq', neuv: 'neuf', un: 'un' };

function ordinalWordToCardinal(word: string): string | null {
  const m = /^(.+?)i[èe]mes?$/.exec(word);
  if (!m) return null;
  const stem = m[1]!;
  const mapped = ORDINAL_STEMS[stem];
  if (mapped) return mapped;
  if (NUMBER_WORDS.has(stem)) return stem;
  if (NUMBER_WORDS.has(`${stem}e`)) return `${stem}e`;
  return null;
}

function parseWords(s: string): string | null {
  const words = splitNumberWords(s);
  if (words.length === 0) return null;
  if (words.length === 1 && /^premi(?:er|ère|ers|ères)$/.test(words[0]!)) return '1e';
  const last = words[words.length - 1]!;
  const ordinalBase = ordinalWordToCardinal(last);
  if (ordinalBase !== null) {
    if (ordinalBase === 'un' && words.length === 1) return null;
    const value = parseCardinal([...words.slice(0, -1), ordinalBase]);
    return value !== null && value > 0 ? `${value}e` : null;
  }
  if (!words.every((x) => NUMBER_WORDS.has(x))) return null;
  const value = parseCardinal(words);
  return value !== null ? String(value) : null;
}

function romanValue(roman: string): number {
  let total = 0;
  for (let i = 0; i < roman.length; i++) {
    const v = ROMAN_VALUES[roman.charAt(i)] ?? 0;
    const next = ROMAN_VALUES[roman.charAt(i + 1)] ?? 0;
    total += v < next ? -v : v;
  }
  return total;
}

function stripLeadingZeros(digits: string): string {
  return digits.replace(/^0+(?=\d)/, '');
}

/** int.frac × 10^zeros as a canonical decimal string. */
function scaleDecimal(intPart: string, fracPart: string, zeros: number): string {
  const frac = fracPart.replace(/0+$/, '');
  if (zeros === 0) return frac.length > 0 ? `${stripLeadingZeros(intPart)}.${frac}` : stripLeadingZeros(intPart);
  const shifted = frac.padEnd(zeros, '0');
  const intDigits = stripLeadingZeros(intPart + shifted.slice(0, zeros));
  const rest = shifted.slice(zeros).replace(/0+$/, '');
  return rest.length > 0 ? `${intDigits}.${rest}` : intDigits;
}

function parseDigits(s: string): string | null {
  const m = DIGITS_RE.exec(s);
  if (!m) return null;
  const [, sign = '', intRaw = '', fracRaw, ordinal, scaleWord] = m;
  const intPart = intRaw.replace(/[ .]/g, '');
  if (ordinal !== undefined && (fracRaw !== undefined || scaleWord !== undefined || sign !== '')) return null;
  const zeros = scaleWord === undefined ? 0 : scaleWord.startsWith('milliard') ? 9 : scaleWord.startsWith('million') ? 6 : 3;
  const value = scaleDecimal(intPart, fracRaw ?? '', zeros);
  if (ordinal !== undefined) return value === '0' ? null : `${value}e`;
  return sign !== '' && sign !== '+' && /[1-9]/.test(value) ? `-${value}` : value;
}

/**
 * Canonical form of a French number token, or null when the token is not a number:
 * « 1 000 », « 1.000 » → "1000"; « 3,5 » → "3.5"; « vingt-et-un » → "21"; « 2,5 millions » → "2500000";
 * ordinals « XIXe », « 19ème », « dix-neuvième » → "19e"; « 1er », « Ire », « première » → "1e".
 * Uppercase roman numerals are read as numbers (except a lone C, D, L or M): callers decide from the context.
 */
export function normalizeNumberFr(token: string): string | null {
  const s = token.normalize('NFKC').replace(/[\u00A0\u202F\u2007]/g, ' ').replace(/\s+/g, ' ').trim();
  if (s.length === 0) return null;
  if (/^[+\-\u2212]?\d/.test(s)) return parseDigits(s);
  const words = parseWords(s);
  if (words !== null) return words;
  const roman = ROMAN_RE.exec(s);
  if (roman) {
    const numeral = s.slice(0, s.length - (roman[5]?.length ?? 0));
    // A lone C, D, L or M is far more often a letter, a title (« M. ») or an article (« Le », « De ») than a number.
    if (/^[CDLM]$/.test(numeral)) return null;
    const value = romanValue(numeral);
    return roman[5] !== undefined ? `${value}e` : String(value);
  }
  return null;
}

File: ios-native/CarnetKit/Sources/CarnetKit/Text/Numbers.swift
import Foundation

/// Reading a French number, however it is written. Ported from `shared/src/text/numbers.ts`.
///
/// « 1 000 », « 1.000 », « mille » and « M » are the same number; « 3,5 » and « 3.5 » too. The app has to know that to
/// check that a model did not change a figure of the book, and to read a page aloud without saying « un virgule cinq »
/// where the page says « 1,5 ».
public enum FrenchNumbers {
    private static let units: [String: Int] = [
        "zéro": 0, "zero": 0, "un": 1, "une": 1, "deux": 2, "trois": 3, "quatre": 4, "cinq": 5, "six": 6, "sept": 7,
        "huit": 8, "neuf": 9, "dix": 10, "onze": 11, "douze": 12, "treize": 13, "quatorze": 14, "quinze": 15, "seize": 16,
    ]
    private static let tens: [String: Int] = [
        "vingt": 20, "vingts": 20, "trente": 30, "quarante": 40, "cinquante": 50, "soixante": 60,
        // Belgian and Swiss forms: the app is read in several countries.
        "septante": 70, "octante": 80, "huitante": 80, "nonante": 90,
    ]
    private static let hundreds: Set<String> = ["cent", "cents"]
    private static let scales: [String: Int] = [
        "mille": 1_000, "mil": 1_000, "million": 1_000_000, "millions": 1_000_000,
        "milliard": 1_000_000_000, "milliards": 1_000_000_000,
    ]

    /// Every word that can appear inside a number written in letters.
    public static let numberWords: Set<String> = {
        var set = Set(units.keys)
        set.formUnion(tens.keys)
        set.formUnion(hundreds)
        set.formUnion(scales.keys)
        return set
    }()

    private static let ordinalSuffix = "(?:e|es|er|ers|re|res|ère|ères|ème|èmes|eme|emes|è|nd|nde|nds|ndes)"
    private static let romanValues: [Character: Int] = ["I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000]

    private static let digitsPattern = try! NSRegularExpression(
        pattern: "^([+\\-\u{2212}]?)(\\d{1,3}(?:[ .]\\d{3})+|\\d+)(?:[,.](\\d+))?(?:(\(ordinalSuffix)))?"
            + "(?: (mille|millions?|milliards?))?$",
        options: []
    )
    private static let romanPattern = try! NSRegularExpression(
        pattern: "^(?=[IVXLCDM])(M{0,3})(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})(\(ordinalSuffix))?$",
        options: []
    )
    private static let ordinalWordPattern = try! NSRegularExpression(pattern: "^(.+?)i[èe]mes?$", options: [])
    private static let firstPattern = try! NSRegularExpression(pattern: "^premi(?:er|ère|ers|ères)$", options: [])

    private static func group(_ match: NSTextCheckingResult, _ index: Int, in text: String) -> String? {
        let range = match.range(at: index)
        guard range.location != NSNotFound, let swiftRange = Range(range, in: text) else { return nil }
        return String(text[swiftRange])
    }

    private static func firstMatch(_ expression: NSRegularExpression, _ text: String) -> NSTextCheckingResult? {
        expression.firstMatch(in: text, options: [], range: NSRange(location: 0, length: text.utf16.count))
    }

    // MARK: - Numbers written in letters

    /// A value and where reading stopped.
    private typealias Parsed = (value: Int, next: Int)

    private static func unit(_ words: [String], _ index: Int, min: Int, max: Int) -> Int? {
        guard index < words.count, let value = units[words[index]], value >= min, value <= max else { return nil }
        return value
    }

    /// 1 to 19 after « soixante » or « quatre-vingt », where the teens keep going; 1 to 9 after the other tens.
    private static func parseTail(_ words: [String], _ index: Int, allowTeens: Bool) -> Parsed? {
        if allowTeens {
            if index < words.count && words[index] == "dix" {
                if let u = unit(words, index + 1, min: 7, max: 9) { return (10 + u, index + 2) }
                return (10, index + 1)
            }
            if let teen = unit(words, index, min: 11, max: 16) { return (teen, index + 1) }
        }
        if let u = unit(words, index, min: 1, max: 9) { return (u, index + 1) }
        return nil
    }

    private static func parseBelow100(_ words: [String], _ index: Int) -> Parsed? {
        guard index < words.count else { return nil }
        let word = words[index]
        let following = index + 1 < words.count ? words[index + 1] : ""

        if word == "quatre" && (following == "vingt" || following == "vingts") {
            let tail = following == "vingt" ? parseTail(words, index + 2, allowTeens: true) : nil
            if let tail { return (80 + tail.value, tail.next) }
            return (80, index + 2)
        }
        if let ten = tens[word] {
            // « soixante-dix » keeps counting into the teens; « trente-dix » does not exist.
            if let tail = parseTail(words, index + 1, allowTeens: ten == 60) { return (ten + tail.value, tail.next) }
            return (ten, index + 1)
        }
        if word == "dix", let u = unit(words, index + 1, min: 7, max: 9) { return (10 + u, index + 2) }
        if let value = units[word] { return (value, index + 1) }
        return nil
    }

    private static func parseBelow1000(_ words: [String], _ index: Int) -> Parsed? {
        var hundredsValue = 0
        var j = index

        if j < words.count && hundreds.contains(words[j]) {
            hundredsValue = 100
            j += 1
        } else if let u = unit(words, j, min: 2, max: 9),
                  j + 1 < words.count, hundreds.contains(words[j + 1]) {
            hundredsValue = u * 100
            j += 2
        }

        if hundredsValue > 0, j < words.count, words[j] == "zéro" || words[j] == "zero" {
            return (hundredsValue, j)
        }
        if let rest = parseBelow100(words, j) {
            if hundredsValue > 0 && rest.value == 0 { return (hundredsValue, j) }
            return (hundredsValue + rest.value, rest.next)
        }
        return hundredsValue > 0 ? (hundredsValue, j) : nil
    }

    /// Value of a whole list of number words, or nil when the list is not exactly one number.
    private static func parseCardinal(_ words: [String]) -> Int? {
        guard !words.isEmpty else { return nil }
        if words.count == 1 && (words[0] == "zéro" || words[0] == "zero") { return 0 }

        var total = 0
        var lastScale = Int.max
        var i = 0
        while i < words.count {
            let group = parseBelow1000(words, i)
            if let group, group.value == 0 { return nil }
            let next = group?.next ?? i
            let scale = next < words.count ? scales[words[next]] : nil

            if let scale {
                // « mille millions » is not a number: the scales must go down.
                if scale >= lastScale { return nil }
                if group == nil && scale != 1_000 { return nil }
                total += (group?.value ?? 1) * scale
                lastScale = scale
                i = next + 1
                continue
            }
            guard let group, next == words.count else { return nil }
            return total + group.value
        }
        return total
    }

    private static func splitNumberWords(_ text: String) -> [String] {
        text.lowercased()
            .split(whereSeparator: { $0.isWhitespace || $0 == "-" || $0 == "\u{2010}" || $0 == "\u{2011}" })
            .map(String.init)
            .filter { !$0.isEmpty && $0 != "et" }
    }

    /// Stems whose ordinal is not simply the cardinal plus « ième ».
    private static let ordinalStems: [String: String] = ["cinqu": "cinq", "neuv": "neuf", "un": "un"]

    private static func ordinalWordToCardinal(_ word: String) -> String? {
        guard let match = firstMatch(ordinalWordPattern, word), let stem = group(match, 1, in: word) else { return nil }
        if let mapped = ordinalStems[stem] { return mapped }
        if numberWords.contains(stem) { return stem }
        if numberWords.contains(stem + "e") { return stem + "e" }
        return nil
    }

    private static func parseWords(_ text: String) -> String? {
        let words = splitNumberWords(text)
        guard !words.isEmpty else { return nil }
        if words.count == 1, firstMatch(firstPattern, words[0]) != nil { return "1e" }

        if let base = ordinalWordToCardinal(words[words.count - 1]) {
            // « unième » alone is not a number; « vingt-et-unième » is.
            if base == "un" && words.count == 1 { return nil }
            let cardinal = Array(words.dropLast()) + [base]
            guard let value = parseCardinal(cardinal), value > 0 else { return nil }
            return "\(value)e"
        }
        guard words.allSatisfy(numberWords.contains) else { return nil }
        guard let value = parseCardinal(words) else { return nil }
        return String(value)
    }

    // MARK: - Roman numerals

    private static func romanValue(_ roman: String) -> Int {
        let letters = Array(roman)
        var total = 0
        for index in letters.indices {
            let value = romanValues[letters[index]] ?? 0
            let next = index + 1 < letters.count ? (romanValues[letters[index + 1]] ?? 0) : 0
            total += value < next ? -value : value
        }
        return total
    }

    // MARK: - Numbers written in digits

    private static func strippedLeadingZeros(_ digits: String) -> String {
        var out = digits
        while out.count > 1 && out.hasPrefix("0") { out.removeFirst() }
        return out
    }

    /// `int.frac` multiplied by ten to the `zeros`, as a canonical decimal string.
    private static func scaleDecimal(_ intPart: String, _ fracPart: String, zeros: Int) -> String {
        var fraction = fracPart
        while fraction.hasSuffix("0") { fraction.removeLast() }
        if zeros == 0 {
            return fraction.isEmpty ? strippedLeadingZeros(intPart) : "\(strippedLeadingZeros(intPart)).\(fraction)"
        }
        let shifted = fraction.padding(toLength: max(zeros, fraction.count), withPad: "0", startingAt: 0)
        let moved = String(shifted.prefix(zeros))
        var rest = String(shifted.dropFirst(zeros))
        while rest.hasSuffix("0") { rest.removeLast() }
        let digits = strippedLeadingZeros(intPart + moved)
        return rest.isEmpty ? digits : "\(digits).\(rest)"
    }

    private static func parseDigits(_ text: String) -> String? {
        guard let match = firstMatch(digitsPattern, text) else { return nil }
        let sign = group(match, 1, in: text) ?? ""
        let intRaw = group(match, 2, in: text) ?? ""
        let fracRaw = group(match, 3, in: text)
        let ordinal = group(match, 4, in: text)
        let scaleWord = group(match, 5, in: text)

        let intPart = intRaw.filter { $0 != " " && $0 != "." }
        // « 3,5e » and « -2e » are not ordinals of anything.
        if ordinal != nil && (fracRaw != nil || scaleWord != nil || !sign.isEmpty) { return nil }

        let zeros: Int
        if let scaleWord {
            zeros = scaleWord.hasPrefix("milliard") ? 9 : (scaleWord.hasPrefix("million") ? 6 : 3)
        } else {
            zeros = 0
        }
        let value = scaleDecimal(intPart, fracRaw ?? "", zeros: zeros)
        if ordinal != nil { return value == "0" ? nil : "\(value)e" }
        let isNegative = !sign.isEmpty && sign != "+" && value.contains(where: { "123456789".contains($0) })
        return isNegative ? "-\(value)" : value
    }

    /// Canonical form of a French number, or nil when the token is not one.
    ///
    /// « 1 000 » and « 1.000 » give "1000"; « 3,5 » gives "3.5"; « vingt-et-un » gives "21";
    /// « 2,5 millions » gives "2500000"; « XIXe », « 19ème » and « dix-neuvième » all give "19e".
    /// A lone C, D, L or M is left alone: far more often a letter, a title (« M. ») or a word than a number.
    public static func normalize(_ token: String) -> String? {
        var text = token.precomposedStringWithCompatibilityMapping
        text = String(text.unicodeScalars.map {
            $0 == "\u{00A0}" || $0 == "\u{202F}" || $0 == "\u{2007}" ? Character(" ") : Character($0)
        })
        text = text.split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
            .trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty else { return nil }

        if let first = text.first, first.isNumber || first == "+" || first == "-" || first == "\u{2212}" {
            return parseDigits(text)
        }
        if let words = parseWords(text) { return words }

        if let match = firstMatch(romanPattern, text) {
            let suffix = group(match, 5, in: text) ?? ""
            let numeral = String(text.dropLast(suffix.count))
            if numeral.count == 1, "CDLM".contains(numeral) { return nil }
            let value = romanValue(numeral)
            return suffix.isEmpty ? String(value) : "\(value)e"
        }
        return nil
    }
}
